### PR TITLE
Fixing bug introduced by PyTorch 1.10

### PR DESCRIPTION
**Pull request by tmarkovich**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

This fixes issue #245.

## How Has This Been Tested (if it applies)

I've confirmed that I've been able to train on PyTorch 1.10 using this fix.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: torchbiggraph/losses.py [hunk 1]
```diff
@@ -147,15 +147,15 @@ def forward(
         if weight is not None:
             loss_per_sample = F.cross_entropy(
                 scores,
-                pos_scores.new_zeros((), dtype=torch.long).expand(num_pos),
+                torch.zeros((num_pos, ), dtype=torch.long, device=scores.device),
```
Comment (adamlerer) [inline torchbiggraph/losses.py:150]:
```suggestion
                pos_scores.new_zeros((num_pos, ), dtype=torch.long),
```

File: torchbiggraph/losses.py [hunk 2]
```diff
@@ -147,15 +147,15 @@ def forward(
...
                 reduction="none",
             )
             match_shape(weight, num_pos)
             loss_per_sample = loss_per_sample * weight
         else:
             loss_per_sample = F.cross_entropy(
                 scores,
-                pos_scores.new_zeros((), dtype=torch.long).expand(num_pos),
+                torch.zeros((num_pos, ), dtype=torch.long, device=scores.device),
```
Comment (adamlerer) [inline torchbiggraph/losses.py:158]:
```suggestion
                pos_scores.new_zeros((num_pos, ), dtype=torch.long),
```

**Review by adamlerer — CHANGES_REQUESTED** [review]
Thanks! Would it be possible to make this small change? I want to confirm that the error is really because of the `expand` rather than because of `new_zeros`.

**Comment by lw** [comment]
This change is undoing a perf optimization: the code before was carefully written to allocate only one element in memory, and then adapt the tensor's indexing so that it _looks like_ that element is replicated many times. The new code however allocates multiple elements in memory. I don't know if the impact is significant, but it would certainly be better to avoid it.

If there has been a regression in PyTorch between 1.9 and 1.10 that should be reported upstream and ideally fixed there, instead of working around it in our code.

**Comment by tmarkovich** [comment]
@lw I agree in principal, but the timeframe on which the upstream bug will be fixed. I've also observed almost no performance degradation in practice with the above change -- with 16 A100 GPUs I'm able to train at 46m edges/second with pytorch 1.9 and `new_zeros` or pytorch 1.10 and the above change.

@adamlerer I did find that the issue is down to the `expand` and not the difference between `zeros` and `new_zeros`.

@lw with that the above in mind, what is the perf difference between `new_zeros((), ...).expand(npos)` and `new_zeros((npos, ))`?

Screen Shot 2022-01-20 at 9 34 47 AM

Anyway, I'll report this bug upstream to pytorch as well.

**Comment by lw** [comment]
> what is the perf difference between new_zeros((), ...).expand(npos) and new_zeros((npos, ))?

It's more of a memory gain (and, secondarily, the time gain to allocate/fill in that memory). But admittedly it's probably barely significant, since if I remember correctly we're talking about allocating 4 extra bytes for each input edge of a batch (rather than just 4 bytes once for the entire batch).

If you've confirmed that the `expand` is indeed the root cause of the bug and that there's no observable difference after this fix, then I'm fine with it. Thanks for reporting the bug upstream anyways!

**Comment by tmarkovich** [comment]
I'll go ahead and change to `new_zeros((npos, ))` and then add an issue to revert the fix once the upstream bug is fixed (with linked issue).

In the mean time, here's the upstream issue:
https://github.com/pytorch/pytorch/issues/71550

**Comment by adamlerer** [comment]
@lw you're right that there's a (minor) performance implication to this, but I think it's overshadowed by the benefit of PyTorch 1.10 compatibility (even if this is fixed in a future version, we ideally want compatibility across a range of pytorch versions).

**Comment by ngimel** [comment]
It is indeed fixed in point release (which everyone should be using instead of 1.10) https://github.com/pytorch/pytorch/pull/69617/files#diff-13bb6989036251a34e2a9f7bd28349761c20950940cd22300a72ffc860296ed0R285, but it's fixed in a way to undo your perf optimization inside cross-entropy implementation, by calling `contiguous()` on your expanded tensor.

**Comment by tmarkovich** [comment]
I just wanted to check in -- it looks like this was imported into the Meta internal repo  -- did it get mainlined externally? If so, should we close this PR? And if not, should I make any changes?

**Comment by adamlerer** [comment]
> It is indeed fixed in point release (which everyone should be using instead of 1.10) https://github.com/pytorch/pytorch/pull/69617/files#diff-13bb6989036251a34e2a9f7bd28349761c20950940cd22300a72ffc860296ed0R285, but it's fixed in a way to undo your perf optimization inside cross-entropy implementation, by calling `contiguous()` on your expanded tensor.

What is "point release" @ngimel ?

**Comment by adamlerer** [comment]
Sorry about the delay @tmarkovich , this PR should merge today.

**Comment by ngimel** [comment]
> What is "point release" @ngimel ?

1.10.1 or 1.10.2 (it's fixed in both).

**Pull request closed without merging**